src/mod_cache_users.erl

```diff
@@ -0,0 +1,268 @@
+%%% @doc Caches info about non-anonymous users using a queue of ets tables
+-module(mod_cache_users).
+
+-behaviour(gen_mod).
+-behaviour(gen_server).
+
+%% gen_server callbacks
+-export([start_link/2]).
+-export([init/1]).
+-export([handle_call/3]).
+-export([handle_cast/2]).
+-export([handle_info/2]).
+
+%% gen_mod API
+-export([start/2]).
+-export([stop/1]).
+-export([config_spec/0]).
+-export([supported_features/0]).
+
+%% Hooks.
+-export([does_cached_user_exist/3]).
+-export([maybe_put_user_into_cache/3]).
+-export([remove_user/3]).
+-export([remove_domain/3]).
+
+-include("mongoose_config_spec.hrl").
+-include("mongoose.hrl").
+-include("jlib.hrl").
+
+tbl_name() ->
+    mod_cache_users.
+
+-spec tbl_name(mongooseim:host_type()) -> atom().
+tbl_name(HostType) ->
+    gen_mod:get_module_proc(HostType, tbl_name()).
+
+%%====================================================================
+%% API
+%%====================================================================
+
+-spec start(mongooseim:host_type(), gen_mod:module_opts()) -> ok.
+start(HostType, Opts) ->
+    start_server(HostType, Opts),
+    ejabberd_hooks:add(hooks(HostType)),
+    ok.
+
+-spec stop(mongooseim:host_type()) -> ok.
+stop(HostType) ->
+    ejabberd_hooks:delete(hooks(HostType)),
+    stop_server(HostType),
+    ok.
+
+-spec config_spec() -> mongoose_config_spec:config_section().
+config_spec() ->
+    #section{items = #{
+                       <<"time_to_live">> => #option{type = int_or_infinity,
+                                                     validate = non_negative,
+                                                     format = {kv, ttl}},
+                       <<"number_of_segments">> => #option{type = integer}
+                      }}.
+
+-spec supported_features() -> [atom()].
+supported_features() ->
+    [dynamic_domains].
+
+hooks(HostType) ->
+    [
+     {does_user_exist, HostType, ?MODULE, does_cached_user_exist, 30},
+     {does_user_exist, HostType, ?MODULE, maybe_put_user_into_cache, 70},
+     {remove_user, HostType, ?MODULE, remove_user, 30},
```

Comment:
We must ensure that `remove_user` and `remove_domain` hooks are executed first for  `ejabberd_auth`, after that for `mod_cache_users` and only then for all other modules.
So there is no chance that we asynchronously re-insert the removed user in the cache, or insert any permanent data (e.g. MAM, Inbox) for the removed but still cached user.
I suggest to change priorities for  `remove_user` & `remove_domain` hooks in the following way:
* ejabberd_auth should have prio=1
* mod_cache_users should have prio=2
* all other modules - prio>30

Reply:
Very good point.

Reply:
In the case of `remove_user`, `mongoose_hooks:remove_user` is only called by `ejabberd_auth:remove_user`, after `ejabberd_auth` has called `remove_user` on all of its backends. So we're safe there.
For `remove_domain`, gonna do that.

Reply:
For `remove_domain`, all modules have priority 50. I've set ejabberd_auth to have 10 and the cache to have 20, just to leave space for other handlers, were any to be desired.